include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,447 @@
+	float scale = exp2(float(exp));//uintBitsToFloat((uint(exp)+127u)<<23u)
+	
+	vec3 v;
+	v.x = int(bitfieldExtract(x.x, 0, RGB19E7_MANTISSA_BITS))*scale;
+	v.y = int(
+		bitfieldExtract(x.x, RGB19E7_MANTISSA_BITS, 32-RGB19E7_MANTISSA_BITS) | 
+		(bitfieldExtract(x.y, 0, RGB19E7_MANTISSA_BITS-(32-RGB19E7_MANTISSA_BITS))<<(32-RGB19E7_MANTISSA_BITS))
+	) * scale;
+	v.z = int(bitfieldExtract(x.y, RGB19E7_MANTISSA_BITS-(32-RGB19E7_MANTISSA_BITS), RGB19E7_MANTISSA_BITS)) * scale;
+	
+	return v;
+}
+
+//i think ill have to create some c++ macro or something to create string with those
+//becasue it's too fucked up to remember about every change in c++ and have to update everything here
+#define INSTR_OPCODE_MASK			0x0fu
+#define INSTR_REG_MASK				0xffu
+#define INSTR_BSDF_BUF_OFFSET_SHIFT 13
+#define INSTR_BSDF_BUF_OFFSET_MASK	0x7ffffu
+#define INSTR_NDF_SHIFT 5
+#define INSTR_NDF_MASK 0x3u
+#define INSTR_ALPHA_U_TEX_SHIFT 4
+#define INSTR_ALPHA_V_TEX_SHIFT 7
+#define INSTR_REFL_TEX_SHIFT 4
+#define INSTR_PLASTIC_REFL_TEX_SHIFT 9
+#define INSTR_TRANS_TEX_SHIFT 4
+#define INSTR_WARD_VARIANT_SHIFT 5
+#define INSTR_WARD_VARIANT_MASK 0x03u
+#define INSTR_FAST_APPROX_SHIFT 5
+#define INSTR_NONLINEAR_SHIFT 8
+#define INSTR_SIGMA_A_TEX_SHIFT 8
+#define INSTR_WEIGHT_TEX_SHIFT 4
+#define INSTR_TWOSIDED_SHIFT 10
+#define INSTR_MASKFLAG_SHIFT 11
+#define INSTR_OPACITY_TEX_SHIFT 12
+#define INSTR_1ST_PARAM_TEX_SHIFT 4
+
+#define INSTR_NORMAL_ID_SHIFT 56u
+#define INSTR_NORMAL_ID_MASK  0xffu
+
+uint instr_getNormalId(in instr_t instr)
+{
+	return (instr.y>>(INSTR_NORMAL_ID_SHIFT-32u)) & INSTR_NORMAL_ID_MASK;
+}
+uint instr_getOpcode(in instr_t instr)
+{
+	return instr.x&INSTR_OPCODE_MASK;
+}
+uint instr_getBSDFbufOffset(in instr_t instr)
+{
+	return (instr.x>>INSTR_BSDF_BUF_OFFSET_SHIFT) & INSTR_BSDF_BUF_OFFSET_MASK;
+}
+uint instr_getNDF(in instr_t instr)
+{
+	return (instr.x>>INSTR_NDF_SHIFT) & INSTR_NDF_MASK;
+}
+bool instr_getAlphaUTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_ALPHA_U_TEX_SHIFT)) != 0u;
+}
+bool instr_getAlphaVTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_ALPHA_V_TEX_SHIFT)) != 0u;
+}
+bool instr_getReflectanceTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_REFL_TEX_SHIFT)) != 0u;
+}
+bool instr_getPlasticReflTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_PLASTIC_REFL_TEX_SHIFT)) != 0u;
+}
+uint instr_getWardVariant(in instr_t instr)
+{
+	return (instr.x>>INSTR_WARD_VARIANT_SHIFT) & INSTR_WARD_VARIANT_MASK;
+}
+bool instr_getFastApprox(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_FAST_APPROX_SHIFT)) != 0u;
+}
+bool instr_getNonlinear(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_NONLINEAR_SHIFT)) != 0u;
+}
+bool instr_getSigmaATexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_SIGMA_A_TEX_SHIFT)) != 0u;
+}
+bool instr_getWeightTexPresence(in instr_t instr)
```

Comment:
you want to conditionally optimize these out if you know that the BSDF forest nodes will always return the same values

Reply:
done?

Reply:
TODO